core/src/multi_proof.rs

```diff
@@ -42,6 +42,106 @@ pub struct MultiProof {
     pub external_siblings: Vec<Node>,
 }
 
+// Given a vector of PathProofs ordered by the key_path,
+// `PathProofRange` represent a range of that vector
+// with key_paths that have common bits up to path_bit_index
+struct PathProofRange {
+    // lower bound of the range
+    lower: usize,
+    // upper bound of the range
+    upper: usize,
+    // bit index in the key path where this struct is pointing at,
+    // this index will be increased or used to create a new bisection
+    path_bit_index: usize,
+}
+
+enum PathProofRangeStep {
+    Bisect {
+        left: PathProofRange,
+        right: PathProofRange,
+    },
+    Advance {
+        sibling: Node,
+    },
+}
+
+impl PathProofRange {
+    fn prove_unique_path_remainder(&self, path_proofs: &Vec<PathProof>) -> Option<SubPathProof> {
+        // If PathProofRange contains only one path_proofs
+        // then all remaining siblings are required for the multiproof
+
+        if self.lower != self.upper - 1 {
+            return None;
```

Comment:
fail fast (panic) if this function requires being called with a single proof range.